File: core-py/klein_core/agent_loop.py
"""!Klein native agent core loop (Python) — ReAct tool-calling on the local model.

Mirrors the TS ``src/agent-core/agent-loop.ts`` (ReAct; arXiv:2210.03629): each turn the model emits a
constrained-JSON action (a tool call or ``final``); the loop runs the tool, feeds back the observation, and
guards against loops (repeated identical action), unknown tools, and a max-turn budget. The action is produced
via :func:`klein_core.structured.generate_structured` so small/quantized models reliably emit a valid choice.

The decider is injectable so the loop is unit-testable without a live model.
"""

from __future__ import annotations

import inspect
import json
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import Any

from .agent_tools import AgentTool
from .generation import ChatMessage, GenerationBackend
from .sampling import resolve_sampling
from .structured import generate_structured


@dataclass
class AgentTranscriptEntry:
    turn: int
    action: dict[str, Any]
    observation: str | None = None
    error: str | None = None


@dataclass
class AgentRunResult:
    status: str  # completed | max_turns | stalled | error
    final_message: str | None
    transcript: list[AgentTranscriptEntry] = field(default_factory=list)
    turns: int = 0


DecideAction = Callable[[str, list[AgentTool], list[AgentTranscriptEntry]], Awaitable[dict[str, Any]]]

_MAX_OBSERVATION_CHARS = 4000


def _stringify(value: Any) -> str:
    if isinstance(value, str):
        text = value
    else:
        try:
            text = json.dumps(value)
        except (TypeError, ValueError):
            text = str(value)
    return text if len(text) <= _MAX_OBSERVATION_CHARS else f"{text[:_MAX_OBSERVATION_CHARS]}\n[truncated]"


def _fingerprint(action: dict[str, Any]) -> str:
    if action.get("action") == "final":
        return "final"
    try:
        return f"{action.get('action')}:{json.dumps(action.get('input'), sort_keys=True)}"
    except (TypeError, ValueError):
        return f"{action.get('action')}:{action.get('input')}"


async def run_agent_loop(
    task: str,
    tools: list[AgentTool],
    decide_action: DecideAction,
    max_turns: int = 20,
    repeated_action_limit: int = 3,
) -> AgentRunResult:
    tools_by_name = {tool.name: tool for tool in tools}
    transcript: list[AgentTranscriptEntry] = []
    last_fingerprint: str | None = None
    repeat_count = 0

    for turn in range(1, max_turns + 1):
        try:
            action = await decide_action(task, tools, transcript)
        except Exception as error:  # noqa: BLE001 - surface decider failure as a terminal status
            transcript.append(AgentTranscriptEntry(turn=turn, action={"action": "final"}, error=str(error)))
            return AgentRunResult(status="error", final_message=None, transcript=transcript, turns=turn)

        if action.get("action") == "final":
            transcript.append(AgentTranscriptEntry(turn=turn, action=action))
            return AgentRunResult(
                status="completed", final_message=str(action.get("message", "")), transcript=transcript, turns=turn
            )

        fingerprint = _fingerprint(action)
        repeat_count = repeat_count + 1 if fingerprint == last_fingerprint else 1
        last_fingerprint = fingerprint
        if repeat_count >= repeated_action_limit:
            transcript.append(
                AgentTranscriptEntry(
                    turn=turn, action=action, error=f"Repeated {action.get('action')} {repeat_count}x; parking."
                )
            )
            return AgentRunResult(status="stalled", final_message=None, transcript=transcript, turns=turn)

        tool = tools_by_name.get(str(action.get("action")))
        entry = AgentTranscriptEntry(turn=turn, action=action)
        if tool is None:
            entry.observation = f"Error: unknown tool {action.get('action')!r}. Tools: {', '.join(tools_by_name)}."
            transcript.append(entry)
            continue
        try:
            result = tool.run(action.get("input") or {})
            if inspect.isawaitable(result):
                result = await result
            entry.observation = _stringify(result)
        except Exception as error:  # noqa: BLE001 - tool errors are observations, not crashes
            entry.observation = f"Error: {error}"
        transcript.append(entry)

    return AgentRunResult(status="max_turns", final_message=None, transcript=transcript, turns=max_turns)


def _action_schema(tools: list[AgentTool]) -> dict[str, Any]:
    return {
        "name": "agent_action",
        "schema": {
            "type": "object",
            "properties": {
                "thought": {"type": "string"},
                "action": {"type": "string", "enum": [*[t.name for t in tools], "final"]},
                "input": {"type": "object"},
                "message": {"type": "string"},
            },
            "required": ["action"],
            "additionalProperties": True,
        },
    }


_SYSTEM_PROMPT = (
    "You are !Klein's local coding agent. Work in small, verifiable steps. Each turn choose exactly one "
    "action: a tool, or 'final'. Reply ONLY with JSON: "
    '{"thought": str, "action": <tool|"final">, "input": object, "message": str}. '
    "Prefer edit_file for changes to existing files."
)


def make_model_decider(backend: GenerationBackend, model_id: str) -> DecideAction:
    """Builds a decider that selects the next action via constrained structured generation."""

    async def decide(task: str, tools: list[AgentTool], transcript: list[AgentTranscriptEntry]) -> dict[str, Any]:
        catalog = "\n".join(f"- {tool.name}: {tool.description}" for tool in tools)
        progress = (
            "\n".join(
                f"Step {e.turn}: {e.action.get('action')} {json.dumps(e.action.get('input'))}"
                + (f" -> {e.observation}" if e.observation else "")
                for e in transcript
            )
            or "(no steps yet)"
        )
        messages = [
            ChatMessage(role="system", content=_SYSTEM_PROMPT),
            ChatMessage(
                role="user",
                content=f"Task:\n{task}\n\nTools:\n{catalog}\n\nProgress:\n{progress}\n\nChoose the next action.",
            ),
        ]
        value = await generate_structured(
            backend,
            model_id,
            messages,
            _action_schema(tools),
            resolve_sampling(role="structured", model_id=model_id),
        )
        return value if isinstance(value, dict) else {"action": "final", "message": str(value)}

    return decide

File: src/agent-core/agent-loop.ts
/**
 * !Klein native agent core — a constrained tool-calling loop that does NOT depend on the NKlein SDK.
 *
 * This is the foundation of !Klein's own backend (the decision to grow beyond a NKlein-only runtime is recorded
 * in THIRD_PARTY_NOTICES.md and plan.md). The loop is the well-trodden ReAct / tool-calling pattern (Yao et al.,
 * ReAct, arXiv:2210.03629), adapted for small/quantized local models: the *next action* is produced via
 * constrained JSON decoding (see `LocalLlmClient.generateStructured`), so the model reliably emits a valid
 * tool selection instead of malformed free-text — the single biggest failure mode for weak models.
 *
 * Design is deliberately decoupled and testable: the loop takes an injected `decideAction` (so it can be unit-
 * tested with a scripted decider) and a list of `AgentCoreTool`s (structurally satisfied by !Klein's existing
 * tools: edit_file, write_files, read, search, repo_map, …). `createLocalLlmActionDecider` wires a real
 * `LocalLlmClient` to `decideAction`.
 */

export interface AgentCoreTool {
	name: string;
	description: string;
	inputSchema?: unknown;
	execute(input: unknown, context?: unknown): Promise<unknown> | unknown;
}

export interface AgentToolAction {
	kind: "tool";
	thought?: string;
	tool: string;
	input: unknown;
}

export interface AgentFinalAction {
	kind: "final";
	thought?: string;
	message: string;
}

export type AgentAction = AgentToolAction | AgentFinalAction;

export interface AgentTranscriptEntry {
	turn: number;
	thought?: string;
	/** The action the model chose this turn. */
	action: AgentAction;
	/** Tool observation (stringified, truncated) when the action was a tool call. */
	observation?: string;
	error?: string;
}

export type AgentRunStatus = "completed" | "max_turns" | "stalled" | "error";

export interface AgentRunResult {
	status: AgentRunStatus;
	finalMessage: string | null;
	transcript: AgentTranscriptEntry[];
	turns: number;
}

export interface DecideActionInput {
	task: string;
	tools: AgentCoreTool[];
	transcript: AgentTranscriptEntry[];
}

export type DecideAction = (input: DecideActionInput) => Promise<AgentAction>;

export interface RunAgentLoopOptions {
	task: string;
	tools: AgentCoreTool[];
	decideAction: DecideAction;
	maxTurns?: number;
	/** Consecutive identical tool actions after which the run is parked as stalled. */
	repeatedActionLimit?: number;
	/** Max characters of a tool observation fed back to the model. */
	maxObservationChars?: number;
	onEntry?: (entry: AgentTranscriptEntry) => void;
}

const DEFAULT_MAX_TURNS = 20;
const DEFAULT_REPEATED_ACTION_LIMIT = 3;
const DEFAULT_MAX_OBSERVATION_CHARS = 4_000;

function stringifyObservation(value: unknown, maxChars: number): string {
	let text: string;
	if (typeof value === "string") {
		text = value;
	} else {
		try {
			text = JSON.stringify(value);
		} catch {
			text = String(value);
		}
	}
	return text.length > maxChars ? `${text.slice(0, maxChars)}\n[truncated]` : text;
}

function actionFingerprint(action: AgentAction): string {
	if (action.kind === "final") {
		return "final";
	}
	let inputKey: string;
	try {
		inputKey = JSON.stringify(action.input);
	} catch {
		inputKey = String(action.input);
	}
	return `${action.tool}:${inputKey}`;
}

export async function runAgentLoop(options: RunAgentLoopOptions): Promise<AgentRunResult> {
	const maxTurns = options.maxTurns ?? DEFAULT_MAX_TURNS;
	const repeatedActionLimit = options.repeatedActionLimit ?? DEFAULT_REPEATED_ACTION_LIMIT;
	const maxObservationChars = options.maxObservationChars ?? DEFAULT_MAX_OBSERVATION_CHARS;
	const toolsByName = new Map(options.tools.map((tool) => [tool.name, tool]));
	const transcript: AgentTranscriptEntry[] = [];

	let lastFingerprint: string | null = null;
	let repeatCount = 0;

	for (let turn = 1; turn <= maxTurns; turn += 1) {
		let action: AgentAction;
		try {
			action = await options.decideAction({ task: options.task, tools: options.tools, transcript });
		} catch (error) {
			const entry: AgentTranscriptEntry = {
				turn,
				action: { kind: "final", message: "" },
				error: error instanceof Error ? error.message : String(error),
			};
			transcript.push(entry);
			options.onEntry?.(entry);
			return { status: "error", finalMessage: null, transcript, turns: turn };
		}

		if (action.kind === "final") {
			const entry: AgentTranscriptEntry = { turn, thought: action.thought, action };
			transcript.push(entry);
			options.onEntry?.(entry);
			return { status: "completed", finalMessage: action.message, transcript, turns: turn };
		}

		// Stall guard: the same tool+input repeated too many times means the model is looping.
		const fingerprint = actionFingerprint(action);
		repeatCount = fingerprint === lastFingerprint ? repeatCount + 1 : 1;
		lastFingerprint = fingerprint;
		if (repeatCount >= repeatedActionLimit) {
			const entry: AgentTranscriptEntry = {
				turn,
				thought: action.thought,
				action,
				error: `Repeated the same ${action.tool} call ${repeatCount} times; parking to avoid a loop.`,
			};
			transcript.push(entry);
			options.onEntry?.(entry);
			return { status: "stalled", finalMessage: null, transcript, turns: turn };
		}

		const tool = toolsByName.get(action.tool);
		const entry: AgentTranscriptEntry = { turn, thought: action.thought, action };
		if (!tool) {
			entry.observation = `Error: unknown tool "${action.tool}". Available tools: ${options.tools
				.map((candidate) => candidate.name)
				.join(", ")}.`;
			transcript.push(entry);
			options.onEntry?.(entry);
			continue;
		}
		try {
			const result = await tool.execute(action.input, { agentId: "nklein-agent-core", iteration: turn });
			entry.observation = stringifyObservation(result, maxObservationChars);
		} catch (error) {
			entry.observation = `Error: ${error instanceof Error ? error.message : String(error)}`;
		}
		transcript.push(entry);
		options.onEntry?.(entry);
	}

	return { status: "max_turns", finalMessage: null, transcript, turns: maxTurns };
}
